src/core/credentials/transport/tls/spiffe_utils.cc

```diff
@@ -156,4 +189,100 @@ absl::StatusOr<SpiffeId> SpiffeId::FromString(absl::string_view input) {
   return SpiffeId(trust_domain, absl::StrCat("/", path));
 }
 
+void SpiffeBundleKey::JsonPostLoad(const Json& json, const JsonArgs&,
+                                   ValidationErrors* errors) {
+  {
+    ValidationErrors::ScopedField field(errors, ".use");
+    if (use_ != kAllowedUse) {
+      errors->AddError(
+          absl::StrFormat("got %s. Only supported value for use field is %s.",
+                          use_, kAllowedUse));
+    }
+  }
+  {
+    ValidationErrors::ScopedField field(errors, ".kty");
+    if (kty_ != kAllowedKty) {
+      errors->AddError(
+          absl::StrFormat("got %s. Only supported value for kty field is %s",
+                          kty_, kAllowedKty));
+    }
+  }
+  {
+    ValidationErrors::ScopedField field(errors, ".x5c");
+    if (x5c_.size() != kx5cSize) {
+      errors->AddError(absl::StrFormat(
+          "got vector length %i. Expected length of exactly %i.", x5c_.size(),
+          kx5cSize));
+    }
+    absl::Status status = ValidateCertificateIsValidX509(x5c_[0]);
+    if (!status.ok()) {
+      errors->AddError(status.message());
+    }
+  }
+}
+
+void SpiffeBundleMap::JsonPostLoad(const Json& json, const JsonArgs&,
+                                   ValidationErrors* errors) {
+  {
+    ValidationErrors::ScopedField field(errors, ".trust_domains");
+    for (auto const& kv : temp_bundles_) {
+      absl::Status status = ValidateTrustDomain(kv.first);
+      if (!status.ok()) {
+        errors->AddError(
+            absl::StrFormat("map key '%s' is not a valid trust domain. %s",
+                            kv.first, status.ToString()));
+      }
+    }
+  }
+
+  // JsonObjectLoader cannot parse into a map with a custom comparator, so parse
+  // into a map without one, then insert those elements into the map with the
+  // custom comparator after parsing. This is a one-time conversion to not have
+  // to create strings every look-up.
+  bundles_.insert(temp_bundles_.begin(), temp_bundles_.end());
+}
+
+absl::StatusOr<SpiffeBundleMap> SpiffeBundleMap::FromFile(
+    absl::string_view file_path) {
+  auto slice = LoadFile(file_path.data(), /*add_null_terminator=*/false);
```

Comment:
Prefer to minimize use of "auto" unless it is evident from context, similarly throughout.